### PR TITLE
Move training_op_helpers to framework

### PR DESCRIPTION
**Pull request by Squadrick**
Continuation of #29703 

Solves #27899

cc: Original reviewers @alextp @jhseu  
cc: Original authors @vcarpani @jhseu for CLAs

### CODE REVIEW

**Comment by Squadrick** [comment]
Any chance this can be CP'd into `2.2`? We're blocked on this for https://github.com/tensorflow/addons/pull/1334

**Comment by vcarpani** [comment]
@googlebot I consent.

File: tensorflow/core/BUILD
Comment (alextp) [inline tensorflow/core/BUILD]: Not public visibility.

We also have many google-internal users of the library in kernels, so we should leave a forwarding header file or this PR will not be successfully merged.

Comment (Squadrick) [reply tensorflow/core/BUILD]: Hey @alextp, I'm not the original author of this PR. I've just checked it out due to it being needed in TensorFlow Addons (original: #29703). I fixed some of the issues during a local compile and created the PR. 

I don't fully understand what you mentioned, could I maybe get a bit more context? More specifically, is there an example in TensorFlow repo or any other that uses `bazel` and makes use of forwarding headers?

Comment (alextp) [reply tensorflow/core/BUILD]: 1. Remove the line saying visibility:public; use at most something like //tensorflow:internal so it doesn't become a free-for-all.

2. Leave a file in tensorflow/core/kernels/training_op_helpers.h which just `#include`s the file in core/framework; this way preexisting code which imports the file in kernels can continue to build while we refactor it to get the library from the right place.

**Comment by Squadrick** [comment]
Most the of CI jobs fail with the same error:
```
Multiple OpKernel registrations match NodeDef at the same priority '{{node Variable}}': 'op: "VariableV2" device_type: "CPU"' and 'op: "VariableV2" device_type: "CPU"'
```

**Review by alextp — DISMISSED** [review]

**Comment by alextp** [comment]
@Squadrick this error means we're violating [C++'s One Definition Rule](https://en.wikipedia.org/wiki/One_Definition_Rule). This is likely happening because two binaries are compiling the same kernel registration. It's likely something to do with the dependency structure of the BUILD files.

Can you try to fix this? I

**Comment by gbaned** [comment]
@Squadrick Can you please check @alextp's comments and keep us posted. Thanks!

File: tensorflow/BUILD
Comment (spidyDev) [inline tensorflow/BUILD:646]: @Squadrick Do we need this target in here. Removing this would get rid of "Multiple OpKernel registrations".

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by Squadrick** [comment]
@tensorflowbutler Still being worked on.

**Comment by gbaned** [comment]
@Squadrick Any update on this PR? and please resolve conflicts Thanks!

**Comment by samikama** [comment]
@Squadrick There is a double inclusion with this PR. I can fix it if you give me write access or I can open a new PR based on this and with fixes. @alextp @jhseu, What do you prefer?

**Comment by Squadrick** [comment]
@samikama Go ahead, I'll close this PR.

**Comment by Squadrick** [comment]
@alextp Sorry about the delay; fixed the errors. There was a small change from your last review, pointed out by @spidyDev.

**Review by alextp — APPROVED** [review]

**Comment by spidyDev** [comment]
@Squadrick  I was trying to use this PR (use the header file in addons) . I keep getting undefined symbol error at runtime. I am not sure if this change is complete .

I just include the header file (#include "tensorflow/core/framework/training_op_helpers.h)  in my /kernels/xxx_ops.cc file
tensorflow.python.framework.errors_impl.NotFoundError: /usr/local/lib/python3.6/dist-packages/tensorflow_addons/custom_ops/optimizers/_optimizer_ops.so: undefined symbol: _ZN10tensorflow7functor11DenseUpdateIN5Eigen9GpuDeviceEdLNS_15DenseUpdateTypeE2EEclERKS3_NS2_9TensorMapINS2_6TensorIdLi1ELi1ElEELi16ENS2_11MakePointerEEENS8_INS9_IKdLi1ELi1ElEELi16ESB_EE

**Comment by Squadrick** [comment]
@spidyDev Have you included the target as a dep in your bazel build file?

**Comment by spidyDev** [comment]
@Squadrick  No , i have not included any target in my bazel file (https://github.com/tensorflow/addons/pull/1334/files).
But custom_op_library() would add  libtensorflow_framework as deps by default .
        "@local_config_tf//:libtensorflow_framework",
        "@local_config_tf//:tf_header_lib",
Please correct me , if I am missing something here.
Also, could you suggest what needs to be added as target. You can add a comment on [PR](https://github.com/tensorflow/addons/pull/1334/files) if that is easier. Thanks !!!

**Comment by samikama** [comment]
@Squadrick, not all custom ops are part of TF tree. TF-Addons for example link against TF library and can't include bazel targets from TF since it is used as external binary. As you proposed above before you closed this PR, I prepared another version which builds fine with TF-Addons custom ops. I will be opening PR now.

**Comment by Squadrick** [comment]
@samikama thanks so much for taking this fixing this. I'll close this PR.

**Pull request closed without merging**
